### PR TITLE
143457/143937 Implement the logic to decrypt the permit string

### PR DESCRIPTION
**Pull request by samad1501673**
1. implemented decryption logic of permit string for new enc cell (PBI:143457)
2. implemented decryption logic of permit string for Update enc enc cell (PBI:143937).
3.  Added Active key and Next key  tags in sap xml.
4. Added UTs.
5. Added functional test case

### CODE REVIEW

File: src/UKHO.ERPFacade.Common/Models/PermitKey.cs [hunk 1]
```diff
@@ -0,0 +1,9 @@
+﻿
+namespace UKHO.ERPFacade.Common.Models
+{
+    public class PermitKey
```
Comment (pravinvarade10801) [inline src/UKHO.ERPFacade.Common/Models/PermitKey.cs:4]: Add ExcludeFromCodeCoverage

File: src/UKHO.ERPFacade.Common/PermitDecryption/PermitDecryption.cs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+                var keys = new PermitKey
+                {
+                    ActiveKey = Convert.ToHexString(firstCellKey),
+                    NextKey = Convert.ToHexString(secondCellKey)
+                };
+                return keys;
+            }
+            catch (Exception ex)
+            {
+                _logger.LogInformation(EventIds.PermitDecryptionException.ToEventId(), ex, "An error occurred while decrypting the permit string.");
```
Comment (pravinvarade10801) [inline src/UKHO.ERPFacade.Common/PermitDecryption/PermitDecryption.cs:43]: Add this as a LogError

File: tests/UKHO.ERPFacade.API.UnitTests/Helpers/EncContentSapMessageBuilderTests.cs [hunk 3]
```diff
@@ -270,11 +434,13 @@ public void SortXmlPayloadTest()
...
+            var actualXmlElement = @"<ACTIONNUMBER>1</ACTIONNUMBER><ACTION>CREATE ENC CELL</ACTION><PRODUCT>ENC CELL</PRODUCT><PRODTYPE>S57</PRODTYPE><CHILDCELL>US5AK83M</CHILDCELL><PRODUCTNAME>US5AK83M</PRODUCTNAME><CANCELLED></CANCELLED><REPLACEDBY></REPLACEDBY><AGENCY>US</AGENCY><PROVIDER>1</PROVIDER><ENCSIZE>small</ENCSIZE><TITLE>St. Michael Bay</TITLE><EDITIONNO>0</EDITIONNO><UPDATENO>1</UPDATENO><UNITTYPE></UNITTYPE><ACTIVEKEY>D9F7832D87</ACTIVEKEY><NEXTKEY>D9F7832D88</NEXTKEY>";
+            var permitKeys = new PermitKey { ActiveKey = "D9F7832D87", NextKey = "D9F7832D88" };
 
             var scenarios = JsonConvert.DeserializeObject<EncEventPayload>(scenariosDataCancelReplaceCell);
             XmlDocument soapXml = new();
             soapXml.LoadXml(sapXmlFile);
+            A.CallTo(() => _fakePermitDecryption.GetPermitKeys(A<string>.Ignored)).Returns(permitKeys);
 
             MethodInfo GetUnitOfSaleForEncCell = typeof(EncContentSapMessageBuilder).GetMethod("GetUnitOfSaleForEncCell", BindingFlags.NonPublic | BindingFlags.Instance)!;
```
Comment (pravinvarade10801) [inline tests/UKHO.ERPFacade.API.UnitTests/Helpers/EncContentSapMessageBuilderTests.cs:445]:
```suggestion
            MethodInfo getUnitOfSaleForEncCell = typeof(EncContentSapMessageBuilder).GetMethod("GetUnitOfSaleForEncCell", BindingFlags.NonPublic | BindingFlags.Instance)!;
```

File: tests/UKHO.ERPFacade.API.UnitTests/Helpers/EncContentSapMessageBuilderTests.cs [hunk 4]
```diff
@@ -25,10 +26,13 @@ public class EncContentSapMessageBuilderTests
         private IFileSystemHelper _fakeFileSystemHelper;
         private IOptions<SapActionConfiguration> _fakeSapActionConfig;
         private ILogger<EncContentSapMessageBuilder> _fakeLogger;
+        private IPermitDecryption _fakePermitDecryption;
 
         private EncContentSapMessageBuilder _fakeEncContentSapMessageBuilder;
         private const string XpathActionItems = $"//*[local-name()='ACTIONITEMS']";
         private const string EncCell = "ENC CELL";
+        private readonly string XpathActiveKey = $"//*[local-name()='ACTIVEKEY']";
```
Comment (pravinvarade10801) [inline tests/UKHO.ERPFacade.API.UnitTests/Helpers/EncContentSapMessageBuilderTests.cs:34]:
```suggestion
        private const string XpathActiveKey = $"//*[local-name()='ACTIVEKEY']";
```

File: tests/UKHO.ERPFacade.API.UnitTests/Helpers/EncContentSapMessageBuilderTests.cs [hunk 5]
```diff
@@ -25,10 +26,13 @@ public class EncContentSapMessageBuilderTests
...
+        private readonly string XpathNextKey = $"//*[local-name()='NEXTKEY']";
```
Comment (pravinvarade10801) [inline tests/UKHO.ERPFacade.API.UnitTests/Helpers/EncContentSapMessageBuilderTests.cs:35]:
```suggestion
        private const string XpathNextKey = $"//*[local-name()='NEXTKEY']";
```

File: tests/UKHO.ERPFacade.API.UnitTests/Helpers/EncContentSapMessageBuilderTests.cs [hunk 6]
```diff
@@ -25,10 +26,13 @@ public class EncContentSapMessageBuilderTests
...
 
         private readonly string sapXmlFile = @"<?xml version=""1.0"" encoding=""utf-8""?>
```
Comment (pravinvarade10801) [inline tests/UKHO.ERPFacade.API.UnitTests/Helpers/EncContentSapMessageBuilderTests.cs:37]: Can we make this variable as constant please. Also do the same for other applicable variables.

File: src/UKHO.ERPFacade.Common/UKHO.ERPFacade.Common.csproj [hunk 7]
```diff
@@ -30,7 +30,9 @@
     <PackageReference Include="Microsoft.Extensions.Options" Version="7.0.1" />
     <PackageReference Include="Azure.Data.Tables" Version="12.8.3" />
     <PackageReference Include="Newtonsoft.Json" Version="13.0.3" />
+    <PackageReference Include="SharpZipLib" Version="1.4.2" />
```
Comment (Vishal14583) [inline src/UKHO.ERPFacade.Common/UKHO.ERPFacade.Common.csproj:33]: Is this package required?

**Review by Vishal14583 — CHANGES_REQUESTED** [review]
Few review comments

File: src/UKHO.ERPFacade.Common/UKHO.ERPFacade.Common.csproj (hunk 7 above)
Comment (samad1501673) [reply src/UKHO.ERPFacade.Common/UKHO.ERPFacade.Common.csproj:33]: Yes

Comment (samad1501673) [reply src/UKHO.ERPFacade.Common/UKHO.ERPFacade.Common.csproj:33]: this package is used in permit decryption logic.

**Review by pravinvarade10801 — APPROVED** [review]
LGTM

File: tests/UKHO.ERPFacade.API.FunctionalTests/appsettings.json [hunk 8]
```diff
@@ -43,6 +43,17 @@
         "SapUsernameForRecordOfSale": "",
         "SapPasswordForRecordOfSale": ""
     },
+    "PermitWithSameKey": {
+        "Permit": "GB4DEVEA202401012120A2629A79A91B2120A2629A79A91B377E9AB7A3180AF4",
```
Comment (Vishal14583) [inline tests/UKHO.ERPFacade.API.FunctionalTests/appsettings.json:47]: Can we add this in keyvault?

**Review by Vishal14583 — CHANGES_REQUESTED** [review]
Use keyvault to store permit keys

**Pull request closed without merging**